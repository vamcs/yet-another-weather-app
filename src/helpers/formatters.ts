export const formatNumber = ({
  value,
  unit,
}: {
  value: number
  unit: 'celsius' | 'millimeter' | 'meter-per-second'
}) => {
  return value.toLocaleString('sv-SE', {
    style: 'unit',
    unit,
  })
}

export const formatTime = ({ time }: { time: string }) => {
  return new Date(time).toLocaleString('sv-SE', {
    timeZone: 'Europe/Stockholm',
  })
}
