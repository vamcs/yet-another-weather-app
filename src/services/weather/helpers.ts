import { Forecast, ForecastOutput } from './types'

export const getCurrentTimeseries = (
  forecast: Forecast
): Forecast['properties']['timeseries'][number] => {
  const now = new Date()

  const currentTimeseries = forecast.properties.timeseries.find(
    (timeseries) => {
      const time = new Date(timeseries.time)

      const isToday = now.getDate() === time.getDate()
      const hourDifference = Math.abs(now.getHours() - time.getHours())
      const differenceFactor = timeseries.data.next_1_hours
        ? 0 // The current hour will match an existing timeseries, so no difference needed.
        : timeseries.data.next_6_hours
          ? 6 // The current hour is within 6 hours of the timeseries
          : 12 // The current hour is within 12 hours of the timeseries

      return isToday && hourDifference <= differenceFactor
    }
  )

  if (!currentTimeseries) {
    throw new Error('No current timeseries found')
  }

  return currentTimeseries
}

export const mapForecast = (forecast: Forecast): ForecastOutput => {
  const currentTimeseries = getCurrentTimeseries(forecast)

  const { instant, next_1_hours, next_6_hours, next_12_hours } =
    currentTimeseries.data

  return {
    current: {
      temperature: instant.details.air_temperature,
      precipitation:
        next_1_hours?.details.precipitation_amount ??
        next_6_hours?.details.precipitation_amount ??
        next_12_hours?.details.precipitation_amount ??
        0,
      wind: instant.details.wind_speed,
      weatherSymbol:
        next_1_hours?.summary.symbol_code ??
        next_6_hours?.summary.symbol_code ??
        next_12_hours?.summary.symbol_code ??
        '',
    },
    timeseries: [], // TODO: Add timeseries
  }
}
