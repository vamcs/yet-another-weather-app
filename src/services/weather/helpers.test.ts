import { getCurrentTimeseries, mapForecast } from './helpers'
import { forecast } from './testFixtures'

describe('getCurrentTimeseries', () => {
  beforeAll(() => {
    jest.useFakeTimers()
  })

  afterAll(() => {
    jest.useRealTimers()
  })

  it('should return the current timeseries', () => {
    jest.setSystemTime(new Date('2025-03-27T11:00:00Z'))

    const currentTimeseries = getCurrentTimeseries(forecast)

    expect(currentTimeseries.time).toEqual('2025-03-27T11:00:00Z')
    expect(currentTimeseries.data.next_1_hours).toBeDefined()
  })

  it("should pick current hour's timeseries", () => {
    jest.setSystemTime(new Date('2025-03-27T15:54:00Z'))

    const currentTimeseries = getCurrentTimeseries(forecast)

    expect(currentTimeseries.time).toEqual('2025-03-27T15:00:00Z')
    expect(currentTimeseries.data.next_1_hours).toBeDefined()
  })

  it('should use next_6_hours if next_1_hours is not available', () => {
    jest.setSystemTime(new Date('2025-03-30T15:54:00Z'))

    const currentTimeseries = getCurrentTimeseries(forecast)

    expect(currentTimeseries.time).toEqual('2025-03-30T12:00:00Z')
    expect(currentTimeseries.data.next_1_hours).toBeUndefined()
    expect(currentTimeseries.data.next_6_hours).toBeDefined()
  })

  it('should throw an error if no timeseries is found', () => {
    jest.setSystemTime(new Date('2025-12-12T15:54:00Z'))

    expect(() => getCurrentTimeseries(forecast)).toThrow(
      'No current timeseries found'
    )
  })
})

describe('mapForecast', () => {
  beforeAll(() => {
    jest.useFakeTimers()
  })

  afterAll(() => {
    jest.useRealTimers()
  })

  it('should map the forecast', () => {
    jest.setSystemTime(new Date('2025-03-27T11:00:00Z'))

    const forecastOutput = mapForecast(forecast)

    expect(forecastOutput.current).toEqual({
      temperature: 8.8,
      precipitation: 0,
      wind: 4.7,
      weatherSymbol: 'cloudy',
    })

    expect(forecastOutput.timeseries).toMatchSnapshot()
  })
})
