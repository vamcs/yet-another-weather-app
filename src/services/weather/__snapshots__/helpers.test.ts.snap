// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapForecast should map the forecast 1`] = `
[
  {
    "precipitation": 0,
    "temperature": 8.8,
    "timeFrom": "2025-03-27T11:00:00Z",
    "weatherSymbol": "cloudy",
    "wind": 4.7,
  },
  {
    "precipitation": 0,
    "temperature": 8,
    "timeFrom": "2025-03-27T12:00:00Z",
    "weatherSymbol": "cloudy",
    "wind": 4.8,
  },
  {
    "precipitation": 0,
    "temperature": 8,
    "timeFrom": "2025-03-27T13:00:00Z",
    "weatherSymbol": "cloudy",
    "wind": 4.4,
  },
  {
    "precipitation": 0,
    "temperature": 8.4,
    "timeFrom": "2025-03-27T14:00:00Z",
    "weatherSymbol": "cloudy",
    "wind": 4.1,
  },
  {
    "precipitation": 0,
    "temperature": 8.7,
    "timeFrom": "2025-03-27T15:00:00Z",
    "weatherSymbol": "cloudy",
    "wind": 4,
  },
  {
    "precipitation": 0,
    "temperature": 9,
    "timeFrom": "2025-03-27T16:00:00Z",
    "weatherSymbol": "cloudy",
    "wind": 4.6,
  },
  {
    "precipitation": 0,
    "temperature": 8.5,
    "timeFrom": "2025-03-27T17:00:00Z",
    "weatherSymbol": "cloudy",
    "wind": 4.9,
  },
  {
    "precipitation": 0,
    "temperature": 8.4,
    "timeFrom": "2025-03-27T18:00:00Z",
    "weatherSymbol": "cloudy",
    "wind": 4.3,
  },
  {
    "precipitation": 0,
    "temperature": 8.3,
    "timeFrom": "2025-03-27T19:00:00Z",
    "weatherSymbol": "cloudy",
    "wind": 4.2,
  },
  {
    "precipitation": 0,
    "temperature": 8.1,
    "timeFrom": "2025-03-27T20:00:00Z",
    "weatherSymbol": "cloudy",
    "wind": 4.5,
  },
  {
    "precipitation": 0,
    "temperature": 8,
    "timeFrom": "2025-03-27T21:00:00Z",
    "weatherSymbol": "cloudy",
    "wind": 4.7,
  },
  {
    "precipitation": 0,
    "temperature": 8,
    "timeFrom": "2025-03-27T22:00:00Z",
    "weatherSymbol": "cloudy",
    "wind": 4.6,
  },
  {
    "precipitation": 0,
    "temperature": 7.9,
    "timeFrom": "2025-03-27T23:00:00Z",
    "weatherSymbol": "cloudy",
    "wind": 4.5,
  },
  {
    "precipitation": 0,
    "temperature": 7.8,
    "timeFrom": "2025-03-28T00:00:00Z",
    "weatherSymbol": "cloudy",
    "wind": 4.5,
  },
  {
    "precipitation": 0,
    "temperature": 7.8,
    "timeFrom": "2025-03-28T01:00:00Z",
    "weatherSymbol": "cloudy",
    "wind": 4.3,
  },
  {
    "precipitation": 0,
    "temperature": 7.8,
    "timeFrom": "2025-03-28T02:00:00Z",
    "weatherSymbol": "cloudy",
    "wind": 4.2,
  },
  {
    "precipitation": 0,
    "temperature": 7.7,
    "timeFrom": "2025-03-28T03:00:00Z",
    "weatherSymbol": "cloudy",
    "wind": 4.2,
  },
  {
    "precipitation": 0,
    "temperature": 7.5,
    "timeFrom": "2025-03-28T04:00:00Z",
    "weatherSymbol": "cloudy",
    "wind": 4.5,
  },
  {
    "precipitation": 0,
    "temperature": 7.7,
    "timeFrom": "2025-03-28T05:00:00Z",
    "weatherSymbol": "cloudy",
    "wind": 4.6,
  },
  {
    "precipitation": 0,
    "temperature": 8,
    "timeFrom": "2025-03-28T06:00:00Z",
    "weatherSymbol": "cloudy",
    "wind": 5,
  },
  {
    "precipitation": 0,
    "temperature": 8.2,
    "timeFrom": "2025-03-28T07:00:00Z",
    "weatherSymbol": "cloudy",
    "wind": 5.3,
  },
  {
    "precipitation": 0,
    "temperature": 8.7,
    "timeFrom": "2025-03-28T08:00:00Z",
    "weatherSymbol": "cloudy",
    "wind": 5.1,
  },
  {
    "precipitation": 0,
    "temperature": 9.6,
    "timeFrom": "2025-03-28T09:00:00Z",
    "weatherSymbol": "cloudy",
    "wind": 4.6,
  },
  {
    "precipitation": 0,
    "temperature": 10.1,
    "timeFrom": "2025-03-28T10:00:00Z",
    "weatherSymbol": "cloudy",
    "wind": 4.6,
  },
  {
    "precipitation": 0,
    "temperature": 11.2,
    "timeFrom": "2025-03-28T11:00:00Z",
    "weatherSymbol": "cloudy",
    "wind": 4.5,
  },
  {
    "precipitation": 0,
    "temperature": 12.5,
    "timeFrom": "2025-03-28T12:00:00Z",
    "weatherSymbol": "partlycloudy_day",
    "wind": 4.5,
  },
  {
    "precipitation": 0,
    "temperature": 13.4,
    "timeFrom": "2025-03-28T13:00:00Z",
    "weatherSymbol": "partlycloudy_day",
    "wind": 4.5,
  },
  {
    "precipitation": 0,
    "temperature": 13.8,
    "timeFrom": "2025-03-28T14:00:00Z",
    "weatherSymbol": "partlycloudy_day",
    "wind": 4.5,
  },
  {
    "precipitation": 0,
    "temperature": 13.7,
    "timeFrom": "2025-03-28T15:00:00Z",
    "weatherSymbol": "partlycloudy_day",
    "wind": 4.3,
  },
  {
    "precipitation": 0,
    "temperature": 13.1,
    "timeFrom": "2025-03-28T16:00:00Z",
    "weatherSymbol": "partlycloudy_day",
    "wind": 3.9,
  },
  {
    "precipitation": 0,
    "temperature": 11.5,
    "timeFrom": "2025-03-28T17:00:00Z",
    "weatherSymbol": "clearsky_night",
    "wind": 3.6,
  },
  {
    "precipitation": 0,
    "temperature": 9.9,
    "timeFrom": "2025-03-28T18:00:00Z",
    "weatherSymbol": "clearsky_night",
    "wind": 3.5,
  },
  {
    "precipitation": 0,
    "temperature": 9.2,
    "timeFrom": "2025-03-28T19:00:00Z",
    "weatherSymbol": "clearsky_night",
    "wind": 3.7,
  },
  {
    "precipitation": 0,
    "temperature": 8.9,
    "timeFrom": "2025-03-28T20:00:00Z",
    "weatherSymbol": "clearsky_night",
    "wind": 4,
  },
  {
    "precipitation": 0,
    "temperature": 8.4,
    "timeFrom": "2025-03-28T21:00:00Z",
    "weatherSymbol": "fair_night",
    "wind": 3.9,
  },
  {
    "precipitation": 0,
    "temperature": 8,
    "timeFrom": "2025-03-28T22:00:00Z",
    "weatherSymbol": "partlycloudy_night",
    "wind": 3.8,
  },
  {
    "precipitation": 0,
    "temperature": 8,
    "timeFrom": "2025-03-28T23:00:00Z",
    "weatherSymbol": "partlycloudy_night",
    "wind": 3.2,
  },
  {
    "precipitation": 0,
    "temperature": 7.5,
    "timeFrom": "2025-03-29T00:00:00Z",
    "weatherSymbol": "clearsky_night",
    "wind": 3.4,
  },
  {
    "precipitation": 0,
    "temperature": 7.2,
    "timeFrom": "2025-03-29T01:00:00Z",
    "weatherSymbol": "clearsky_night",
    "wind": 3.4,
  },
  {
    "precipitation": 0,
    "temperature": 7,
    "timeFrom": "2025-03-29T02:00:00Z",
    "weatherSymbol": "clearsky_night",
    "wind": 3.3,
  },
  {
    "precipitation": 0,
    "temperature": 6.8,
    "timeFrom": "2025-03-29T03:00:00Z",
    "weatherSymbol": "clearsky_night",
    "wind": 3.3,
  },
  {
    "precipitation": 0,
    "temperature": 6.7,
    "timeFrom": "2025-03-29T04:00:00Z",
    "weatherSymbol": "clearsky_day",
    "wind": 3.3,
  },
  {
    "precipitation": 0,
    "temperature": 6.6,
    "timeFrom": "2025-03-29T05:00:00Z",
    "weatherSymbol": "clearsky_day",
    "wind": 3.2,
  },
  {
    "precipitation": 0,
    "temperature": 7,
    "timeFrom": "2025-03-29T06:00:00Z",
    "weatherSymbol": "fair_day",
    "wind": 3.2,
  },
  {
    "precipitation": 0,
    "temperature": 8.1,
    "timeFrom": "2025-03-29T07:00:00Z",
    "weatherSymbol": "partlycloudy_day",
    "wind": 3.4,
  },
  {
    "precipitation": 0,
    "temperature": 8.9,
    "timeFrom": "2025-03-29T08:00:00Z",
    "weatherSymbol": "partlycloudy_day",
    "wind": 3,
  },
  {
    "precipitation": 0,
    "temperature": 9.7,
    "timeFrom": "2025-03-29T09:00:00Z",
    "weatherSymbol": "cloudy",
    "wind": 3.5,
  },
  {
    "precipitation": 0,
    "temperature": 10.3,
    "timeFrom": "2025-03-29T10:00:00Z",
    "weatherSymbol": "cloudy",
    "wind": 3.4,
  },
  {
    "precipitation": 0,
    "temperature": 11.5,
    "timeFrom": "2025-03-29T11:00:00Z",
    "weatherSymbol": "cloudy",
    "wind": 3.8,
  },
  {
    "precipitation": 0,
    "temperature": 12.3,
    "timeFrom": "2025-03-29T12:00:00Z",
    "weatherSymbol": "cloudy",
    "wind": 3.7,
  },
  {
    "precipitation": 0,
    "temperature": 12.6,
    "timeFrom": "2025-03-29T13:00:00Z",
    "weatherSymbol": "cloudy",
    "wind": 3.3,
  },
  {
    "precipitation": 0,
    "temperature": 13.1,
    "timeFrom": "2025-03-29T14:00:00Z",
    "weatherSymbol": "cloudy",
    "wind": 3.3,
  },
  {
    "precipitation": 0,
    "temperature": 12.9,
    "timeFrom": "2025-03-29T15:00:00Z",
    "weatherSymbol": "cloudy",
    "wind": 3.2,
  },
  {
    "precipitation": 0,
    "temperature": 12.3,
    "timeFrom": "2025-03-29T16:00:00Z",
    "weatherSymbol": "cloudy",
    "wind": 3.2,
  },
  {
    "precipitation": 0,
    "temperature": 11.3,
    "timeFrom": "2025-03-29T17:00:00Z",
    "weatherSymbol": "cloudy",
    "wind": 3.2,
  },
  {
    "precipitation": 0,
    "temperature": 10.1,
    "timeFrom": "2025-03-29T18:00:00Z",
    "timeTo": "2025-03-30T00:00:00Z",
    "weatherSymbol": "cloudy",
    "wind": 3.4,
  },
  {
    "precipitation": 0,
    "temperature": 7.1,
    "timeFrom": "2025-03-30T00:00:00Z",
    "timeTo": "2025-03-30T06:00:00Z",
    "weatherSymbol": "clearsky_night",
    "wind": 2.6,
  },
  {
    "precipitation": 0,
    "temperature": 4.2,
    "timeFrom": "2025-03-30T06:00:00Z",
    "timeTo": "2025-03-30T12:00:00Z",
    "weatherSymbol": "cloudy",
    "wind": 2.5,
  },
  {
    "precipitation": 0,
    "temperature": 8.8,
    "timeFrom": "2025-03-30T12:00:00Z",
    "timeTo": "2025-03-30T18:00:00Z",
    "weatherSymbol": "cloudy",
    "wind": 2.9,
  },
  {
    "precipitation": 0,
    "temperature": 7.2,
    "timeFrom": "2025-03-30T18:00:00Z",
    "timeTo": "2025-03-31T00:00:00Z",
    "weatherSymbol": "cloudy",
    "wind": 1.6,
  },
  {
    "precipitation": 0,
    "temperature": 5.2,
    "timeFrom": "2025-03-31T00:00:00Z",
    "timeTo": "2025-03-31T06:00:00Z",
    "weatherSymbol": "cloudy",
    "wind": 2.1,
  },
  {
    "precipitation": 0,
    "temperature": 3.8,
    "timeFrom": "2025-03-31T06:00:00Z",
    "timeTo": "2025-03-31T12:00:00Z",
    "weatherSymbol": "cloudy",
    "wind": 2.9,
  },
  {
    "precipitation": 0,
    "temperature": 9.3,
    "timeFrom": "2025-03-31T12:00:00Z",
    "timeTo": "2025-03-31T18:00:00Z",
    "weatherSymbol": "cloudy",
    "wind": 4,
  },
  {
    "precipitation": 0,
    "temperature": 8,
    "timeFrom": "2025-03-31T18:00:00Z",
    "timeTo": "2025-04-01T00:00:00Z",
    "weatherSymbol": "partlycloudy_night",
    "wind": 2.3,
  },
  {
    "precipitation": 0,
    "temperature": 5.5,
    "timeFrom": "2025-04-01T00:00:00Z",
    "timeTo": "2025-04-01T06:00:00Z",
    "weatherSymbol": "fair_night",
    "wind": 2.8,
  },
  {
    "precipitation": 0,
    "temperature": 4.8,
    "timeFrom": "2025-04-01T06:00:00Z",
    "timeTo": "2025-04-01T12:00:00Z",
    "weatherSymbol": "fair_day",
    "wind": 3.1,
  },
  {
    "precipitation": 0,
    "temperature": 13.1,
    "timeFrom": "2025-04-01T12:00:00Z",
    "timeTo": "2025-04-01T18:00:00Z",
    "weatherSymbol": "partlycloudy_day",
    "wind": 4.1,
  },
  {
    "precipitation": 0,
    "temperature": 11.2,
    "timeFrom": "2025-04-01T18:00:00Z",
    "timeTo": "2025-04-02T00:00:00Z",
    "weatherSymbol": "fair_night",
    "wind": 3,
  },
  {
    "precipitation": 0,
    "temperature": 8.2,
    "timeFrom": "2025-04-02T00:00:00Z",
    "timeTo": "2025-04-02T06:00:00Z",
    "weatherSymbol": "fair_night",
    "wind": 3.1,
  },
  {
    "precipitation": 0,
    "temperature": 6,
    "timeFrom": "2025-04-02T06:00:00Z",
    "timeTo": "2025-04-02T12:00:00Z",
    "weatherSymbol": "fair_day",
    "wind": 2.9,
  },
  {
    "precipitation": 0,
    "temperature": 13.2,
    "timeFrom": "2025-04-02T12:00:00Z",
    "timeTo": "2025-04-02T18:00:00Z",
    "weatherSymbol": "partlycloudy_day",
    "wind": 3.7,
  },
  {
    "precipitation": 0,
    "temperature": 11.1,
    "timeFrom": "2025-04-02T18:00:00Z",
    "timeTo": "2025-04-03T00:00:00Z",
    "weatherSymbol": "partlycloudy_night",
    "wind": 2.1,
  },
  {
    "precipitation": 0,
    "temperature": 7.7,
    "timeFrom": "2025-04-03T00:00:00Z",
    "timeTo": "2025-04-03T06:00:00Z",
    "weatherSymbol": "partlycloudy_night",
    "wind": 2.2,
  },
  {
    "precipitation": 0,
    "temperature": 6.6,
    "timeFrom": "2025-04-03T06:00:00Z",
    "timeTo": "2025-04-03T12:00:00Z",
    "weatherSymbol": "cloudy",
    "wind": 2.5,
  },
  {
    "precipitation": 0,
    "temperature": 15.9,
    "timeFrom": "2025-04-03T12:00:00Z",
    "timeTo": "2025-04-03T18:00:00Z",
    "weatherSymbol": "fair_day",
    "wind": 3.4,
  },
  {
    "precipitation": 0,
    "temperature": 12,
    "timeFrom": "2025-04-03T18:00:00Z",
    "timeTo": "2025-04-04T00:00:00Z",
    "weatherSymbol": "fair_night",
    "wind": 3.1,
  },
  {
    "precipitation": 0,
    "temperature": 8,
    "timeFrom": "2025-04-04T00:00:00Z",
    "timeTo": "2025-04-04T06:00:00Z",
    "weatherSymbol": "fair_night",
    "wind": 3.4,
  },
  {
    "precipitation": 0,
    "temperature": 5.8,
    "timeFrom": "2025-04-04T06:00:00Z",
    "timeTo": "2025-04-04T12:00:00Z",
    "weatherSymbol": "fair_day",
    "wind": 3.6,
  },
  {
    "precipitation": 0,
    "temperature": 12.7,
    "timeFrom": "2025-04-04T12:00:00Z",
    "timeTo": "2025-04-04T18:00:00Z",
    "weatherSymbol": "cloudy",
    "wind": 4.8,
  },
  {
    "precipitation": 0,
    "temperature": 7.8,
    "timeFrom": "2025-04-04T18:00:00Z",
    "timeTo": "2025-04-05T00:00:00Z",
    "weatherSymbol": "cloudy",
    "wind": 3.6,
  },
  {
    "precipitation": 0,
    "temperature": 5.6,
    "timeFrom": "2025-04-05T00:00:00Z",
    "timeTo": "2025-04-05T06:00:00Z",
    "weatherSymbol": "fair_night",
    "wind": 4,
  },
  {
    "precipitation": 0,
    "temperature": 3,
    "timeFrom": "2025-04-05T06:00:00Z",
    "timeTo": "2025-04-05T12:00:00Z",
    "weatherSymbol": "clearsky_day",
    "wind": 4.7,
  },
  {
    "precipitation": 0,
    "temperature": 7.1,
    "timeFrom": "2025-04-05T12:00:00Z",
    "timeTo": "2025-04-05T18:00:00Z",
    "weatherSymbol": "partlycloudy_day",
    "wind": 5.3,
  },
  {
    "precipitation": 0,
    "temperature": 6,
    "timeFrom": "2025-04-05T18:00:00Z",
    "timeTo": "2025-04-06T00:00:00Z",
    "weatherSymbol": "cloudy",
    "wind": 3.4,
  },
  {
    "precipitation": 0,
    "temperature": 3.2,
    "timeFrom": "2025-04-06T00:00:00Z",
    "timeTo": undefined,
    "weatherSymbol": "clearsky_night",
    "wind": 3.7,
  },
]
`;
